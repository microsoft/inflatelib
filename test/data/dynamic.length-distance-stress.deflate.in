
# This is a "stress" test in that it tests all length/distance codes as well as their minimum and maximums. This gives
# a total of 49 lengths to test (29 codes + 20 with extra bits) and 56 distances (30 + 26 with extra bits). We begin
# with the bytes 0-255, and in order to make the data more predictable, we start by testing all possible lengths, each
# with a distance of 256. This ensures that we repeat the sequence 0-255 for all of the length tests (which will make up
# a majority of the output). For the distances, we have no control over the data it copies from, so we use small lengths
# to make the expected output shorter and easier to produce.
# NOTE: This test is only applicable to Deflate
>1
1       # bfinal = true
10      # Compressed with dynamic codes

11101   # HLIT = 286 (29 + 257)
11101   # HDIST = 30 (29 + 1)
1110    # HCLEN = 18 (14 + 4)

# Code Length Alphabet Code Lengths:
001 000 000 000 100 101 100 011 000 011 000 000 000 101 000 101 000 101

# Literal/Length & Distance Alphabet Code Lengths:
>>1
1100
0 >1 00 >>1
0 >1 11 >>1
0 >1 01 >>1
1101
0 >1 00 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 11 >>1
0 >1 10 >>1
11110
1101
0 >1 11 >>1
1100
0 >1 01 >>1
0 >1 11 >>1
11111
1100
0 >1 00 >>1
0 >1 01 >>1
11101
101
0 >1 00 >>1
100
100
100
101
0 >1 00 >>1
0 >1 01 >>1
11100
101
0 >1 00 >>1
0 >1 11 >>1
0 >1 01 >>1

# Literal/Length Tree:
#   Symbol      Bit Count   Code
#   0           8           01100010
#   1           8           01100011
#   2           8           01100100
#   3           8           01100101
#   4           8           01100110
#   5           8           01100111
#   6           8           01101000
#   7           8           01101001
#   8           8           01101010
#   9           8           01101011
#   10          8           01101100
#   11          8           01101101
#   12          8           01101110
#   13          8           01101111
#   14          9           100000110
#   15          9           100000111
#   16          9           100001000
#   17          9           100001001
#   18          9           100001010
#   19          9           100001011
#   20          9           100001100
#   21          9           100001101
#   22          9           100001110
#   23          9           100001111
#   24          9           100010000
#   25          9           100010001
#   26          9           100010010
#   27          9           100010011
#   28          9           100010100
#   29          9           100010101
#   30          9           100010110
#   31          9           100010111
#   ' '         9           100011000
#   '!'         9           100011001
#   '"'         9           100011010
#   '#'         9           100011011
#   '$'         9           100011100
#   '%'         9           100011101
#   '&'         9           100011110
#   '''         9           100011111
#   '('         9           100100000
#   ')'         9           100100001
#   '*'         9           100100010
#   '+'         9           100100011
#   ','         9           100100100
#   '-'         9           100100101
#   '.'         9           100100110
#   '/'         9           100100111
#   '0'         9           100101000
#   '1'         9           100101001
#   '2'         9           100101010
#   '3'         9           100101011
#   '4'         9           100101100
#   '5'         9           100101101
#   '6'         9           100101110
#   '7'         9           100101111
#   '8'         9           100110000
#   '9'         9           100110001
#   ':'         9           100110010
#   ';'         9           100110011
#   '<'         9           100110100
#   '='         9           100110101
#   '>'         9           100110110
#   '?'         9           100110111
#   '@'         9           100111000
#   'A'         9           100111001
#   'B'         9           100111010
#   'C'         9           100111011
#   'D'         9           100111100
#   'E'         9           100111101
#   'F'         9           100111110
#   'G'         9           100111111
#   'H'         9           101000000
#   'I'         9           101000001
#   'J'         9           101000010
#   'K'         9           101000011
#   'L'         9           101000100
#   'M'         9           101000101
#   'N'         9           101000110
#   'O'         9           101000111
#   'P'         9           101001000
#   'Q'         9           101001001
#   'R'         9           101001010
#   'S'         9           101001011
#   'T'         9           101001100
#   'U'         9           101001101
#   'V'         9           101001110
#   'W'         9           101001111
#   'X'         9           101010000
#   'Y'         9           101010001
#   'Z'         9           101010010
#   '['         9           101010011
#   '\'         9           101010100
#   ']'         9           101010101
#   '^'         9           101010110
#   '_'         9           101010111
#   '`'         9           101011000
#   'a'         9           101011001
#   'b'         9           101011010
#   'c'         9           101011011
#   'd'         9           101011100
#   'e'         9           101011101
#   'f'         9           101011110
#   'g'         9           101011111
#   'h'         9           101100000
#   'i'         9           101100001
#   'j'         9           101100010
#   'k'         9           101100011
#   'l'         9           101100100
#   'm'         9           101100101
#   'n'         9           101100110
#   'o'         9           101100111
#   'p'         9           101101000
#   'q'         9           101101001
#   'r'         9           101101010
#   's'         9           101101011
#   't'         9           101101100
#   'u'         9           101101101
#   'v'         9           101101110
#   'w'         9           101101111
#   'x'         9           101110000
#   'y'         9           101110001
#   'z'         9           101110010
#   '{'         9           101110011
#   '|'         9           101110100
#   '}'         9           101110101
#   '~'         9           101110110
#   127         9           101110111
#   128         9           101111000
#   129         9           101111001
#   130         9           101111010
#   131         9           101111011
#   132         9           101111100
#   133         9           101111101
#   134         9           101111110
#   135         9           101111111
#   136         9           110000000
#   137         9           110000001
#   138         9           110000010
#   139         9           110000011
#   140         9           110000100
#   141         9           110000101
#   142         9           110000110
#   143         9           110000111
#   144         9           110001000
#   145         9           110001001
#   146         9           110001010
#   147         9           110001011
#   148         9           110001100
#   149         9           110001101
#   150         9           110001110
#   151         9           110001111
#   152         9           110010000
#   153         9           110010001
#   154         9           110010010
#   155         9           110010011
#   156         9           110010100
#   157         9           110010101
#   158         9           110010110
#   159         9           110010111
#   160         9           110011000
#   161         9           110011001
#   162         9           110011010
#   163         9           110011011
#   164         9           110011100
#   165         9           110011101
#   166         9           110011110
#   167         9           110011111
#   168         9           110100000
#   169         9           110100001
#   170         9           110100010
#   171         9           110100011
#   172         9           110100100
#   173         9           110100101
#   174         9           110100110
#   175         9           110100111
#   176         9           110101000
#   177         9           110101001
#   178         9           110101010
#   179         9           110101011
#   180         9           110101100
#   181         9           110101101
#   182         9           110101110
#   183         9           110101111
#   184         9           110110000
#   185         9           110110001
#   186         9           110110010
#   187         9           110110011
#   188         9           110110100
#   189         9           110110101
#   190         9           110110110
#   191         9           110110111
#   192         9           110111000
#   193         9           110111001
#   194         9           110111010
#   195         9           110111011
#   196         9           110111100
#   197         9           110111101
#   198         9           110111110
#   199         9           110111111
#   200         9           111000000
#   201         9           111000001
#   202         9           111000010
#   203         9           111000011
#   204         9           111000100
#   205         9           111000101
#   206         9           111000110
#   207         9           111000111
#   208         9           111001000
#   209         9           111001001
#   210         9           111001010
#   211         9           111001011
#   212         9           111001100
#   213         9           111001101
#   214         9           111001110
#   215         9           111001111
#   216         9           111010000
#   217         9           111010001
#   218         9           111010010
#   219         9           111010011
#   220         9           111010100
#   221         9           111010101
#   222         9           111010110
#   223         9           111010111
#   224         9           111011000
#   225         9           111011001
#   226         9           111011010
#   227         9           111011011
#   228         9           111011100
#   229         9           111011101
#   230         9           111011110
#   231         9           111011111
#   232         9           111100000
#   233         9           111100001
#   234         9           111100010
#   235         9           111100011
#   236         9           111100100
#   237         9           111100101
#   238         9           111100110
#   239         9           111100111
#   240         9           111101000
#   241         9           111101001
#   242         9           111101010
#   243         9           111101011
#   244         9           111101100
#   245         9           111101101
#   246         9           111101110
#   247         9           111101111
#   248         9           111110000
#   249         9           111110001
#   250         9           111110010
#   251         9           111110011
#   252         9           111110100
#   253         9           111110101
#   254         9           111110110
#   255         9           111110111
#   END         9           111111000
#   257         3           010
#   258         9           111111001
#   259         9           111111010
#   260         9           111111011
#   261         9           111111100
#   262         9           111111101
#   263         9           111111110
#   264         9           111111111
#   265         8           01110000
#   266         8           01110001
#   267         8           01110010
#   268         8           01110011
#   269         8           01110100
#   270         8           01110101
#   271         8           01110110
#   272         8           01110111
#   273         8           01111000
#   274         8           01111001
#   275         8           01111010
#   276         7           0110000
#   277         8           01111011
#   278         8           01111100
#   279         8           01111101
#   280         8           01111110
#   281         8           01111111
#   282         8           10000000
#   283         8           10000001
#   284         8           10000010
#   285         2           00
#
# Distance Tree:
#   Symbol      Bit Count   Code
#   0           6           100110
#   1           6           100111
#   2           6           101000
#   3           6           101001
#   4           5           10000
#   5           5           10001
#   6           5           10010
#   7           6           101010
#   8           6           101011
#   9           6           101100
#   10          6           101101
#   11          6           101110
#   12          6           101111
#   13          6           110000
#   14          6           110001
#   15          1           0
#   16          6           110010
#   17          6           110011
#   18          6           110100
#   19          6           110101
#   20          6           110110
#   21          6           110111
#   22          6           111000
#   23          6           111001
#   24          6           111010
#   25          6           111011
#   26          6           111100
#   27          6           111101
#   28          6           111110
#   29          6           111111

# Encoded Data:
>>1

# First step: write the bytes 0-255
01100010
01100011
01100100
01100101
01100110
01100111
01101000
01101001
01101010
01101011
01101100
01101101
01101110
01101111
100000110
100000111
100001000
100001001
100001010
100001011
100001100
100001101
100001110
100001111
100010000
100010001
100010010
100010011
100010100
100010101
100010110
100010111
100011000
100011001
100011010
100011011
100011100
100011101
100011110
100011111
100100000
100100001
100100010
100100011
100100100
100100101
100100110
100100111
100101000
100101001
100101010
100101011
100101100
100101101
100101110
100101111
100110000
100110001
100110010
100110011
100110100
100110101
100110110
100110111
100111000
100111001
100111010
100111011
100111100
100111101
100111110
100111111
101000000
101000001
101000010
101000011
101000100
101000101
101000110
101000111
101001000
101001001
101001010
101001011
101001100
101001101
101001110
101001111
101010000
101010001
101010010
101010011
101010100
101010101
101010110
101010111
101011000
101011001
101011010
101011011
101011100
101011101
101011110
101011111
101100000
101100001
101100010
101100011
101100100
101100101
101100110
101100111
101101000
101101001
101101010
101101011
101101100
101101101
101101110
101101111
101110000
101110001
101110010
101110011
101110100
101110101
101110110
101110111
101111000
101111001
101111010
101111011
101111100
101111101
101111110
101111111
110000000
110000001
110000010
110000011
110000100
110000101
110000110
110000111
110001000
110001001
110001010
110001011
110001100
110001101
110001110
110001111
110010000
110010001
110010010
110010011
110010100
110010101
110010110
110010111
110011000
110011001
110011010
110011011
110011100
110011101
110011110
110011111
110100000
110100001
110100010
110100011
110100100
110100101
110100110
110100111
110101000
110101001
110101010
110101011
110101100
110101101
110101110
110101111
110110000
110110001
110110010
110110011
110110100
110110101
110110110
110110111
110111000
110111001
110111010
110111011
110111100
110111101
110111110
110111111
111000000
111000001
111000010
111000011
111000100
111000101
111000110
111000111
111001000
111001001
111001010
111001011
111001100
111001101
111001110
111001111
111010000
111010001
111010010
111010011
111010100
111010101
111010110
111010111
111011000
111011001
111011010
111011011
111011100
111011101
111011110
111011111
111100000
111100001
111100010
111100011
111100100
111100101
111100110
111100111
111101000
111101001
111101010
111101011
111101100
111101101
111101110
111101111
111110000
111110001
111110010
111110011
111110100
111110101
111110110
111110111

# Second step: exhaust all the lengths. This will write 3385 bytes (3641 bytes in total), which means we will have
# written the entire sequence 0-255 14 times, followed by the sequence 0-56
010 0 >1 111111 >>1
111111001 0 >1 111111 >>1
111111010 0 >1 111111 >>1
111111011 0 >1 111111 >>1
111111100 0 >1 111111 >>1
111111101 0 >1 111111 >>1
111111110 0 >1 111111 >>1
111111111 0 >1 111111 >>1
01110000 >1 0 >>1 0 >1 111111 >>1
01110000 >1 1 >>1 0 >1 111111 >>1
01110001 >1 0 >>1 0 >1 111111 >>1
01110001 >1 1 >>1 0 >1 111111 >>1
01110010 >1 0 >>1 0 >1 111111 >>1
01110010 >1 1 >>1 0 >1 111111 >>1
01110011 >1 0 >>1 0 >1 111111 >>1
01110011 >1 1 >>1 0 >1 111111 >>1
01110100 >1 00 >>1 0 >1 111111 >>1
01110100 >1 11 >>1 0 >1 111111 >>1
01110101 >1 00 >>1 0 >1 111111 >>1
01110101 >1 11 >>1 0 >1 111111 >>1
01110110 >1 00 >>1 0 >1 111111 >>1
01110110 >1 11 >>1 0 >1 111111 >>1
01110111 >1 00 >>1 0 >1 111111 >>1
01110111 >1 11 >>1 0 >1 111111 >>1
01111000 >1 000 >>1 0 >1 111111 >>1
01111000 >1 111 >>1 0 >1 111111 >>1
01111001 >1 000 >>1 0 >1 111111 >>1
01111001 >1 111 >>1 0 >1 111111 >>1
01111010 >1 000 >>1 0 >1 111111 >>1
01111010 >1 111 >>1 0 >1 111111 >>1
0110000 >1 000 >>1 0 >1 111111 >>1
0110000 >1 111 >>1 0 >1 111111 >>1
01111011 >1 0000 >>1 0 >1 111111 >>1
01111011 >1 1111 >>1 0 >1 111111 >>1
01111100 >1 0000 >>1 0 >1 111111 >>1
01111100 >1 1111 >>1 0 >1 111111 >>1
01111101 >1 0000 >>1 0 >1 111111 >>1
01111101 >1 1111 >>1 0 >1 111111 >>1
01111110 >1 0000 >>1 0 >1 111111 >>1
01111110 >1 1111 >>1 0 >1 111111 >>1
01111111 >1 00000 >>1 0 >1 111111 >>1
01111111 >1 11111 >>1 0 >1 111111 >>1
10000000 >1 00000 >>1 0 >1 111111 >>1
10000000 >1 11111 >>1 0 >1 111111 >>1
10000001 >1 00000 >>1 0 >1 111111 >>1
10000001 >1 11111 >>1 0 >1 111111 >>1
10000010 >1 00000 >>1 0 >1 111111 >>1
10000010 >1 11110 >>1 0 >1 111111 >>1
00 0 >1 111111 >>1

# Third step: before we test all of the distances, we need to make sure that we have enough data in the distances window
# to actually test all possible distances. At this point in time, there are 3,641 bytes in the window. By the time we
# test the largest distance, we need 32,768 bytes in the buffer. There are 55 copies as a part of the distance testing
# before that point in time, each of which will copy 3 bytes (for a total of 165 bytes copied). That means that we need
# to first write an additional 28,962 bytes. We do this by repeating the 0-255 sequence
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
00 0 >1 111111 >>1
0110000 >1 111 >>1 0 >1 111111 >>1

# Final step: exhaust all the distances. The output here is not as easy to calculate and is done by hand
010 100110
010 100111
010 101000
010 101001
010 10000 >1 0 >>1
010 10000 >1 1 >>1
010 10001 >1 0 >>1
010 10001 >1 1 >>1
010 10010 >1 00 >>1
010 10010 >1 11 >>1
010 101010 >1 00 >>1
010 101010 >1 11 >>1
010 101011 >1 000 >>1
010 101011 >1 111 >>1
010 101100 >1 000 >>1
010 101100 >1 111 >>1
010 101101 >1 0000 >>1
010 101101 >1 1111 >>1
010 101110 >1 0000 >>1
010 101110 >1 1111 >>1
010 101111 >1 00000 >>1
010 101111 >1 11111 >>1
010 110000 >1 00000 >>1
010 110000 >1 11111 >>1
010 110001 >1 000000 >>1
010 110001 >1 111111 >>1
010 0 >1 000000 >>1
010 0 >1 111111 >>1
010 110010 >1 0000000 >>1
010 110010 >1 1111111 >>1
010 110011 >1 0000000 >>1
010 110011 >1 1111111 >>1
010 110100 >1 00000000 >>1
010 110100 >1 11111111 >>1
010 110101 >1 00000000 >>1
010 110101 >1 11111111 >>1
010 110110 >1 000000000 >>1
010 110110 >1 111111111 >>1
010 110111 >1 000000000 >>1
010 110111 >1 111111111 >>1
010 111000 >1 0000000000 >>1
010 111000 >1 1111111111 >>1
010 111001 >1 0000000000 >>1
010 111001 >1 1111111111 >>1
010 111010 >1 00000000000 >>1
010 111010 >1 11111111111 >>1
010 111011 >1 00000000000 >>1
010 111011 >1 11111111111 >>1
010 111100 >1 000000000000 >>1
010 111100 >1 111111111111 >>1
010 111101 >1 000000000000 >>1
010 111101 >1 111111111111 >>1
010 111110 >1 0000000000000 >>1
010 111110 >1 1111111111111 >>1
010 111111 >1 0000000000000 >>1
010 111111 >1 1111111111111 >>1

# And finally, the end
111111000
