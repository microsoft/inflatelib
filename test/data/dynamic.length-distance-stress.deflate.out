
# Data repeated by all of the length codes plus the extra data we need to write to ensure enough data for the distance test
<dec
repeat (127) { 0...255 }
0...90

repeat (57) { 90 }
84...86
86...88
58...60
60...62
32...34
34...36
230...232
232...234
172...174
174...176
50...52
52...54
184...186
186...188
190...192
192...194
196...198
198...200
202...204
204...206
208...210
210...212
214...216
216...218
220...222
222...224
226...228
228...230
232...234
234...236
238...240
240...242
244...246
246...248
250...252
252...254
0...2
